### PR TITLE
detect inplace modifications of views earlier (fix #21875)

### LINKED ISSUE #21875: In-place operation on differentiable view leaks memory
```python
import torch
import gc

gc.collect()
print(torch.cuda.memory_allocated(0))  # prints 0

m = torch.nn.Linear(2,3)
mw = m.weight[:]
m.to('cuda')
# with torch.no_grad():  # Without using `torch.no_grad()`, the next line will leak CUDA memory
mw[0][0] = 5

del mw
del m
gc.collect()
print(torch.cuda.memory_allocated(0))  # prints 512
```

This is reproducible on master.

Initial investigation shows that the leak is not because of saving Variables for backward (the SavedVariable constructor was not called).

cc @ezyang @SsnL @albanD @zou3519 @gqchen @ngimel

### PR DESCRIPTION
**Pull request by bdhirsh**
Updating a check to detect in-place modifications of views of leaf tensors that are opted into autograd. See #21875 for more details, but doing so causes memory leaks.

Note that this is technically a BC-breaking change. Previously, performing an in-place op on a view of a leaf tensor with autograd enabled would not fail- now it does. e.g.

```
a = torch.zeros(2, requires_grad=True)
b = a[1]
b.add_(1) # wouldn't fail unless you called b.backward() before, now this line itself fails
```

Previously, it would fail with a more cryptic error when you made a call to `.backward()`, so this behavior is easier to reason about. Also even though it was raising an error, the graph that created was not valid and lead to a memory leak. The new error prevents this leak from happening.
The previous behavior can be recovered by making the original Tensor not require gradients (as you don't call backward on it anyways).

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46204 detect inplace modifications of views earlier (fix #21875)**

Differential Revision: [D24259500](https://our.internmc.facebook.com/intern/diff/D24259500)

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -3482,6 +3482,13 @@ def test():
         test()
         self.assertEqual(dealloc[0], 1)
 
+    def test_inplace_view_leaf_errors(self):
+        # Issue #21875: Fail faster (when we try to modify the view vs. in backward())
+        x = torch.zeros(1, requires_grad=True)
+        y = x
```
Comment (albanD) [inline test/test_autograd.py:3488]: This is not creating a view :D 
Just assigning the python variable named `y` to point to the same thing as the python variable `x`.
So both `x` and `y` point to the same thing here.

If you want to create a view, you need to do something with the Tensor like: `y = x.view_as(x)`.

File: test/test_autograd.py [hunk 2]
```diff
@@ -3482,6 +3482,13 @@ def test():
...
+        with self.assertRaises(RuntimeError):
```
Comment (albanD) [inline test/test_autograd.py:3489]: You can use `assertRaisesRegex()` here with the content of the error message. That way you can be sure that the right error is raised.
And I would expect this test to fail without the view fix above.
```suggestion
        with self.assertRaisesRegex(RuntimeError, "a view of a leaf Variable that requires grad"):
```

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 3]
```diff
@@ -47,6 +47,11 @@ inline void check_inplace(const Tensor& tensor) {
       auto diff_view_meta = static_cast<DifferentiableViewMeta*>(impl::get_autograd_meta(var));
       // This can throw or warn
       handle_view_on_rebase(diff_view_meta);
+      auto base = tensor._base();
```
Comment (albanD) [inline torch/csrc/autograd/VariableTypeUtils.h:50]: nit: no need to create a temporary Tensor here.
And if you do so, you can take a const reference to avoid the refcount bumps.

**Review by albanD — COMMENTED** [review]
Looks quite good.
I'm not sure the test is doing what you want though :D

File: test/test_autograd.py (hunk 1 above)
Comment (bdhirsh) [reply test/test_autograd.py:3488]: Got it, thanks! :)

**Review by albanD — APPROVED** [review]
The lint needs to be fixed.
But otherwise LGTM.

File: test/test_autograd.py [hunk 4]
```diff
@@ -3536,13 +3536,10 @@ def test_inplace_view_weak_grad_fn(self):
...
 
-        # Issue 23502: Ensure RuntimeError for modification of SavedVariable.
+        # Issue #21875: Fail faster (when we try to modify the view vs. in backward())
         a = torch.rand(10, requires_grad=True).narrow(0, 0, 10)
-        b = a.relu_()
-        c = b.add_(100)
-        del b
         with self.assertRaises(RuntimeError):
-            c.sum().backward(torch.ones(1, requires_grad=True))
+            b = a.relu_()
```
Comment (albanD) [inline test/test_autograd.py:3542]: This is great!

**Comment by gchanan** [comment]
On the BC note -- it's not just that it is easier to reason about, it also prevents a memory leak, right?

**Comment by albanD** [comment]
Updated it to mention that and provide a way to migrate code to avoid the error for code that used to work.
